File: Problem 1/scope.js
/*
 * scope.js
 * This code was written to demonstrate how variable scope works in javascript.
 * In writing this code, I consulted various pages of the MDN javascript docs,
 * which I explicitly cite below. 
 *
 * Docs:
 * 	Closures https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures
 * 	Var https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/var
 *	a re-introduction to javascript
 *	https://developer.mozilla.org/en-US/docs/Web/JavaScript/A_re-introduction_to_JavaScript
 *
 * Nick Creel - Apr 11, 2020 - MIT License
 */
let x = 1 // this is global and can be modified within the same scope
var y = 2 // this is global and can be modified within the same scope 
const z = 3 // this is global, but cannot be modified. 

function testScope(){
	if(x===1){
                let x = 4 // this does not change the value of the global to 4
		// but creates a new local variable called x that contains the 
		// value 4. if we had not used the let keyword in front of this
		// assignment, then the global variable would be changed. Unlike
		// var, declaring a local variable with the same name as
		// a global variable declared with let is possible. 
		console.log(x) // this should print 4.
	}else {
		console.log("value of x has changed")				
	}
	if(y===2){
		y = 5 // this changes the value of the global
		// note that we do not have to use something like the global 
		// statement in python to do this; this is just a feature of
		// how var is declared. if we attempted to declare var y = 5,
		// then var would be declared undefined both globally and locally
		// because var declarations are handled before any code is executed.
		// this creates a conflict between the var declared here and the 
		// var declared globally. 
	}else{
		console.log("value of y has changed")
	}

	var w = "this variable was declared inside of testScope()" 
	// because this is declared within a function scope,
	// this variable is contained within the function 
	// and not accessible outside of the local scope.
	console.log(x) // even though we created a new x, it is not accessible outside
	//              of the if statement in which it was defined.
	// 		this is different from python, which allows variables defined
	// 		within if statements to be used outside of those if statements. 
	console.log(y)
	console.log(z)
}

console.log(x)
console.log(y) // var does not change globally until testScope is run
console.log(z)
// console.log(w) if we ran this, we'd get a reference error, meaning w has not yet been named. 
// even though var is handled before the code is run, var declared within a function block 
// is still confined to that function. 

testScope()

console.log(x)
console.log(y)
console.log(z)
// console.log(w) if we ran this, we'd still get an error, even after running testScope, for
// the same reasons described above. 

// I will also demonstrate a closure, although I admit I have not seen closures before until
// this exam. My understanding of closures comes primarily from the MDN docs. I cite the 
// appropriate documentation at the top of this file. In any case, I wrote a closure similar
// to the closure in that documentation, though slightly different. 
// i discuss closures in more detail in problem1.txt
function legalName(last) {
	return function(first){
		return first + last
	}
}

var mahoney = legalName("Mahoney")
console.log(mahoney("jim")) // should print jimMahoney

File: Problem 1/scope.py
"""
scope.py

the code in this file is meant to demonstrate how variable scope 
works in python 3.8. When writing this code, I referred to the Python 3.8
documentation to ensure that I was using the correct technical vocabulary
in my own documentation, but all code written is my own. 

Doc pages I referred to:
    Execution Model https://docs.python.org/3.8/reference/executionmodel.html
    Objects, Values, and Types https://docs.python.org/3.8/reference/datamodel.html

Nick Creel - Apr 11, 2020 - MIT License
"""
# the scope of a code block or variable is determined when the block/var is named
from baz import Baz
# because Baz is imported at the module level, it is globally defined. 
bar = 2 
# bar is a global variable because it is defined within module block
# all function and class definitions at the module level are also global
class Baz():
    # because this class is defined after we imported another class named Baz,
    # this particular definition will be resolved instead of the imported Baz. 
    # see the scopeTest method and the output of this file when run to confirm.
    def __init__(self):
        self.bar = 3 
        #not only is this bar local, but can only be accessed as attr of a named
        #instance of this class, because it is attached to the self keyword. 
        #in addition, this variable can be changed in a particular instance,
        #but that will not change the value of the variable in the class defintion.
        # variables declared using the "self." prefix is known as
        # an instance attribute. 
    def scopeTest(self): 
        # this method is a class attribute.
        # this means that it is shared across all instances of the class.
        # unlike an instance attribute, a class attribute may be called 
        # by calling through the class (Baz().scopeTest()) or by calling
        # through an instance (baz = Baz(), baz.scopeTest()). 
        # the defintion of a class instance may be changed at an instance
        # level (so, a particular instance could change the function of scopeTest),
        # but the original class definition would remain unaltered. 
        # even though this method may be called without creating an instance,
        # it cannot be called as a function without referencing the class
        # for example, simply calling scopeTest() without Baz(). or an instance
        # would result in a NameError. 
        print("inside scopeTest function scope")
        print("printing bar")
        print(bar)
        
        # because there's a variable named bar in the nearest enclosing scope, 
        # this will always print the global bar rather than, say, a local bar
        # defined at the same scope as an instance of this class. 
        # see the foo() fxn for an example.
        # HOWEVER, see the following for an example where the global would
        # not be resolved: 
        # bar = 5 
        # if the above line were uncommented, there would be an error, namely
        # an UnboundLocalError, as the print(bar) statement above would be called
        # before bar was assigned in local scope. In this case, the executed code 
        # cannot just resolve the global bar, as the local bar is within the same
        # scope, and bar is not preceded by a global or nonlocal statement. . 
        print("printing self.bar")
        print(self.bar)

def foo():
    baz = Baz()
    print("executing foo()")
    bar = 1 # this is a local variable bound to the function scope
            # this variable may only be used within function scope,
            # and it does not reassign the global bar variable. 
    print(bar)
    
    baz.bar = 6
    print("printing baz.scopeTesi() inside foo()")
    baz.scopeTest()
    print("printing Baz().scopeTest() inside foo()")
    Baz().scopeTest()
    print("exiting foo() scope")
    
foo() # should print 1
print("printing bar outside of foo()")
print(bar) # should print 2
